openedx/adg/lms/applications/api/views.py

```diff
@@ -144,3 +145,56 @@ def update_is_work_experience_not_applicable(self, request):
                 user_application.applications_workexperiences.all().delete()
 
         return Response({}, status=HTTP_200_OK)
+
+
+class ReferenceViewSet(ApplicationRequirementsViewSet):
+    """
+    ViewSet for reference APIs
+
+    retrieve:
+        Return a single reference record that matches requested id.
+
+    **Example Requests**
+    GET /api/applications/reference/
+
+    list:
+        Return all reference records belong to the requested user.
+
+    **Example Requests**
+    GET /api/applications/reference/
+
+    create:
+        Create a new reference record.
+
+    Following points must be kept in mind when creating a new reference record:
+
+    * Id field is read only and would have no effect if provided explicitly.
+
+    **Example Requests**
+    POST /api/applications/reference/
+
+    delete:
+        Remove an existing reference record.
+
+    **Example Requests**
+    DELETE /api/applications/reference/
+
+    update:
+        Update an existing reference record.
+
+    **Example Requests**
+    PUT /api/applications/reference/
+    """
+
+    queryset = Reference.objects.all()
+    serializer_class = ReferenceSerializer
+
+    def create(self, request, *args, **kwargs):
+        """
+        Extending `create` action to prevent exceeding max reference limit against a user application
+        """
+        application_id = request.data['user_application']
+        if Reference.objects.filter(user_application=application_id).count() == MAX_NUMBER_OF_REFERENCES:
+            raise serializers.ValidationError(MAX_REFERENCE_ERROR_MSG)
```

Comment:
This validation is intentionally added here. If we add a `validate_user_application` method in the serializer class, it will perform the validation for PUT requests as well, and won't allow updating a reference of a user application if two references already exist against that user application. We specifically want the validation to be applied against the `create` action, i.e. a POST request.